imap_processing/ena_maps/ena_maps.py

```diff
@@ -0,0 +1,536 @@
+    """
+
+    PUSH = "Push"
+    PULL = "Pull"
+
+
+# Define the pointing set classes
+class PointingSet(ABC):
+    """
+    Abstract class to contain pointing set (PSET) data in the context of ENA maps.
+
+    Parameters
+    ----------
+    dataset : xr.Dataset
+        Dataset containing the pointing set data.
+    reference_frame : geometry.SpiceFrame
+        The reference Spice frame of the pointing set.
+    """
+
+    @abstractmethod
+    def __init__(self, dataset: xr.Dataset, reference_frame: geometry.SpiceFrame):
+        self.reference_frame = reference_frame
+        self.num_points = 0
+        self.az_el_points = np.zeros((self.num_points, 2))
+        self.data = xr.Dataset()
+
+    @abstractmethod
+    def project_to_frame(
+        self, out_frame: geometry.SpiceFrame, event_time: float | None = None
+    ) -> None:
+        """
+        Project the pointing set to a new frame (placeholder).
+
+        Parameters
+        ----------
+        out_frame : geometry.SpiceFrame
+            The frame to which the pointing set should be projected.
+        event_time : float
+            The event time at which to project the pointing set into the new frame.
+        """
+        pass
+
+    def __repr__(self) -> str:
+        """
+        Return a string representation of the pointing set.
+
+        Returns
+        -------
+        str
+            String representation of the pointing set.
+        """
+        return f"{self.__class__} PointingSet(reference_frame={self.reference_frame})"
+
+
+class UltraPointingSet(PointingSet):
+    """
+    PSET object specifically for ULTRA data, nominally at Level 1C.
+
+    Parameters
+    ----------
+    l1c_dataset : xr.Dataset | pathlib.Path | str
+        L1c xarray dataset containing the pointing set data or the path to the dataset.
+        Currently, the dataset is expected to be in a rectangular grid,
+        with data_vars indexed along the coordinates:
+            - 'epoch' : time value (1 value per PSET)
+            - 'azimuth_bin_center' : azimuth bin center values
+            - 'elevation_bin_center' : elevation bin center values
+        Some data_vars may additionally be indexed by energy bin;
+        however, only the spatial axes are used in this class.
+    reference_frame : geometry.SpiceFrame
+        The reference Spice frame of the pointing set. Default is IMAP_DPS.
+    order : {'C', 'F'}, optional
+        The order of the grid to be used in any raveling processes.
+        Default is 'F' (Fortran-style ordering).
+
+    Raises
+    ------
+    ValueError
+        If the azimuth or elevation bin centers do not match the constructed grid.
+        Or if the azimuth or elevation bin spacing is not uniform.
+    ValueError
+        If multiple epochs are found in the dataset.
+    """
+
+    def __init__(
+        self,
+        l1c_dataset: xr.Dataset | pathlib.Path | str,
+        reference_frame: geometry.SpiceFrame = geometry.SpiceFrame.IMAP_DPS,
```

Comment:
Call this `pset_frame` to be explicit? I was confused whether the reference was the map or the pointing set.

Reply:
Done - would you also want the `Map.reference_frame` to be `Map.map_frame`, or is it clear with this change?

Reply:
Internally, they're both stored as `spice_reference_frame` to work interchangeably with the index matching function